File: tests/fixtures/timeseries/verify_johansen_diff.R
# verify_johansen_diff.R - urca 侧 Johansen 双库 diff 网格 (M3 前置, §6.2.1)
#
# 冻结结论见 docs/phases/phase7/JOHANSEN_DUAL_LIB_DIFF.md:
#   - 参数映射: urca ca.jo(K) ≙ statsmodels coint_johansen(k_ar_diff = K-1)
#   - 情形映射: ecdet="none" ≡ SM det_order=0; ecdet="const"/"trend" 无 SM 对应
#   - spec transitory/longrun 数学恒等 (差被 Z1 张成), 仅转储 transitory
#
# 网格: K ∈ {2, 3} × ecdet ∈ {none, const, trend} (transitory)
# 输出: johansen_urca_grid.txt (teststat/cval 统一翻转为 r=0..N-1 与 SM 一致)

.libPaths(c("F:/R/win-library/4.6", .libPaths()))
library(urca)

HERE <- "F:/Cpp_Hub/tests/fixtures/timeseries"
y <- as.matrix(read.csv(file.path(HERE, "coint_smoke_data.csv")))

emit <- function(fh, tag, obj) {
  a <- as.numeric(as.matrix(obj))
  write(sprintf("%s\t%s", tag, paste(sprintf("%.17g", a), collapse = " ")),
        file = fh, append = TRUE)
}

out <- file.path(HERE, "johansen_urca_grid.txt")
fh <- file(out, open = "wt")
for (K in c(2, 3)) {
  for (ec in c("none", "const", "trend")) {
    tag <- sprintf("ur_K%d_%s", K, ec)
    jtr <- ca.jo(y, type = "trace", ecdet = ec, K = K, spec = "transitory")
    jmx <- ca.jo(y, type = "eigen", ecdet = ec, K = K, spec = "transitory")
    # urca 原始序 [r<=N-1, ..., r=0] → 翻转为 [r=0, ..., r<=N-1] (SM 行序)
    emit(fh, paste0(tag, "_eig"), jtr@lambda)
    emit(fh, paste0(tag, "_lr1"), rev(jtr@teststat))
    emit(fh, paste0(tag, "_lr2"), rev(jmx@teststat))
    # cval 行序同 teststat (行1=r<=N-1) → 翻转行序: 行1=r=0
    emit(fh, paste0(tag, "_cvt"), jtr@cval[nrow(jtr@cval):1, ])
    emit(fh, paste0(tag, "_cvm"), jmx@cval[nrow(jmx@cval):1, ])
    emit(fh, paste0(tag, "_nobs"), nrow(jtr@Z0))
  }
}
close(fh)
cat("wrote", out, "\n")

File: tests/fixtures/timeseries/verify_johansen_diff.py
# -*- coding: utf-8 -*-
# verify_johansen_diff.py - statsmodels 侧 Johansen 双库 diff 网格 (M3 前置, §6.2.1)
#
# 冻结结论见 docs/phases/phase7/JOHANSEN_DUAL_LIB_DIFF.md:
#   - 参数映射: urca ca.jo(K) ≙ coint_johansen(k_ar_diff = K-1)
#   - 情形映射: SM det_order=0 ≡ urca ecdet="none" (短回归无约束常数); det -1/1 无 urca 对应
#   - urca spec 参数数学恒等 (y_{t-1}-y_{t-K} ∈ span(Z1))
#
# 网格: det_order ∈ {-1, 0, 1} × k_ar_diff ∈ {1, 2} → johansen_sm_grid.txt
# 配对分析: probe_johansen_pair.py (读双 grid, 输出匹配对)

import os
import numpy as np
from statsmodels.tsa.vector_ar.vecm import coint_johansen

HERE = os.path.dirname(os.path.abspath(__file__))


def emit(fh, tag, obj):
    a = np.asarray(obj, dtype=float).ravel()
    fh.write("%s\t%s\n" % (tag, " ".join("%.17g" % v for v in a)))


def main():
    y = np.loadtxt(os.path.join(HERE, "coint_smoke_data.csv"),
                   delimiter=",", skiprows=1)
    out = os.path.join(HERE, "johansen_sm_grid.txt")
    with open(out, "w") as fh:
        for det in (-1, 0, 1):
            for k in (1, 2):
                tag = "sm_det%d_k%d" % (det, k)
                r = coint_johansen(y, det, k)
                emit(fh, tag + "_eig", r.eig)
                emit(fh, tag + "_lr1", r.lr1)   # trace, r=0..N-1
                emit(fh, tag + "_lr2", r.lr2)   # maxeig, r=0..N-1
                emit(fh, tag + "_cvt", r.cvt)   # N×3 行序 r=0..N-1 (行=rank, 列=90/95/99)
                emit(fh, tag + "_cvm", r.cvm)
                emit(fh, tag + "_evec", r.evec)
                fh.write("%s_nobs\t%d\n" % (tag, r.rkt.shape[0]))
    print("wrote", out)


if __name__ == "__main__":
    main()
